### PR TITLE
Dimensions of vector subspaces

### PR DESCRIPTION
**Pull request by martinra**

### CODE REVIEW

**Comment by thofma** [comment]
@wbhart Does this look good?

**Comment by wbhart** [comment]
@martinra This needs to have some randomised test code. Consider using rand_module (over QQ) as we have done with other test functions, then take a submodule of it.

The documentation (submodule.md) file also needs to actually refer to the docstring, otherwise it does not appear in the docs. (Bonus marks for an example of its use.)

**Comment by martinra** [comment]
I'll take care of this soon, but not quite sure I'll manage this week. Maybe.

**Comment by martinra** [comment]
May I ask you to have a look at the documentation? I don't manage to let the Documenter detect the new dim functions, even though the tests pass without any problem.

File: docs/src/quotient_module.md [hunk 1]
```diff
@@ -90,6 +90,8 @@ following functionality.
 
 ```@docs
 supermodule(M::AbstractAlgebra.Generic.QuotientModule{T}) where T <: RingElement
+
+dim(N::AbstractAlgebra.Generic.Submodule{T}) where T <: FieldElement
```
Comment (thofma) [inline docs/src/quotient_module.md]: `QuotientModule`

File: docs/src/quotient_module.md [hunk 2]
```diff
@@ -118,5 +120,10 @@ Codomain: Quotient module over Integers with 2 generators and relations:
 julia> supermodule(Q) == M
 true
 
+julia> dim(M)
+
```
Comment (thofma) [inline docs/src/quotient_module.md]: Missing the return value.

**Comment by thofma** [comment]
You included two times the same docstring (for `Submodule`). Documenter (the package for assembling the documentation) does not like it. This could be the reason.

**Comment by martinra** [comment]
Thanks for the super quick reply and sorry for this slip in the documentation. The error that I mentioned persists, though, and occurred also before I added dim for quotient spaces. Here is the head of the log for Documenter in a fresh environment. I can't seem to figure it out.

```
(v1.1) pkg> activate AbsAlg
[ Info: activating new environment at ~/.local-oscar/AbstractAlgebra.jl/AbsAlg.

(AbsAlg) pkg> add Documenter
[  Updating registry at `~/.julia/registries/General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
 d Resolving package versions...
d Abstract  Updating `~/.local-oscar/AbstractAlgebra.jl/AbsAlg/Project.toml`
  [e30172f5] + Documenter v0.23.3
  Updating `~/.local-oscar/AbstractAlgebra.jl/AbsAlg/Manifest.toml`
  [ffbed154] + DocStringExtensions v0.8.0
  [e30172f5] + Documenter v0.23.3
  [682c06a0] + JSON v0.21.0
  [69de0a69] + Parsers v0.3.7
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8ba89e20] + Distributed
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode

(AbsAlg) pkg> dev --local .
 Resolving package versions...
  Updating `~/.local-oscar/AbstractAlgebra.jl/AbsAlg/Project.toml`
  [c3fe647b] + AbstractAlgebra v0.6.0 [`..`]
  Updating `~/.local-oscar/AbstractAlgebra.jl/AbsAlg/Manifest.toml`
  [c3fe647b] + AbstractAlgebra v0.6.0 [`..`]
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [2f01184e] + SparseArrays

julia> include("docs/make.jl")
[ Info: Recompiling stale cache file /home/martin/.julia/compiled/v1.1/AbstractAlgebra/b8V2b.ji for AbstractAlgebra [c3fe647b-3220-5bb0-a1ea-a7954cac585d]
[ Info: SetupBuildDirectory: setting up build directory.
[ Info: Doctest: running doctests.
Map with the following data

Domain:
=======
Integers

Codomain:
========
IntegersMap with the following data

Domain:
=======
Integers

Codomain:
========
Integers┌ Error: doctest failure in src/submodule.md:102-140
│
│ ```jldoctest
│ julia> M = FreeModule(ZZ, 2)
│ Free module of rank 2 over Integers
│
│ julia> m = M([ZZ(2), ZZ(3)])
│ (2, 3)
│
│ julia> n = M([ZZ(1), ZZ(4)])
│ (1, 4)
│
│ julia> N1, = sub(M, [m, n])
│ (Submodule over Integers with 2 generators and no relations
│ , Module homomorphism with
│ Domain: Submodule over Integers with 2 generators and no relations
│
│ Codomain: Free module of rank 2 over Integers)
│
│ julia> N2, = sub(M, [m])
│ (Submodule over Integers with 1 generator and no relations
│ , Module homomorphism with
│ Domain: Submodule over Integers with 1 generator and no relations
│
│ Codomain: Free module of rank 2 over Integers)
│
│ julia> supermodule(N1) == M
│ true
│
│ julia> iscompatible(N1, N2)
│ (true, Free module of rank 2 over Integers)
│
│ julia> issubmodule(N1, M)
│ false
│
│ julia> dim(M)
│ 2
│
│ julia> dim(N1)
│ 1
│
│ ```
│
│ Subexpression:
│
│ dim(M)
│
│ Evaluated output:
│
│ ERROR: MethodError: no method matching dim(::AbstractAlgebra.Generic.FreeModule{BigInt})
│ Closest candidates are:
│   dim(!Matched::AbstractAlgebra.Generic.YoungTableau) at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:736
│   dim(!Matched::Type{T<:Integer}, !Matched::AbstractAlgebra.Generic.YoungTableau) where T<:Integer at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:739
│   dim(!Matched::AbstractAlgebra.Generic.FreeModule{T<:Union{FieldElem, AbstractFloat, Rational}}) where T<:Union{FieldElem, AbstractFloat, Rational} at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/FreeModule.jl:44
│   ...
│ Stacktrace:
│  [1] top-level scope at none:0
│
│ Expected output:
│
│ 2
│
│   diff =
│    2ERROR: MethodError: no method matching dim(::AbstractAlgebra.Generic.FreeModule{BigInt})
│    Closest candidates are:
│      dim(!Matched::AbstractAlgebra.Generic.YoungTableau) at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:736
│      dim(!Matched::Type{T<:Integer}, !Matched::AbstractAlgebra.Generic.YoungTableau) where T<:Integer at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:739
│      dim(!Matched::AbstractAlgebra.Generic.FreeModule{T<:Union{FieldElem, AbstractFloat, Rational}}) where T<:Union{FieldElem, AbstractFloat, Rational} at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/FreeModule.jl:44
│      ...
│    Stacktrace:
│     [1] top-level scope at none:0
└ @ Documenter.DocTests ~/.julia/packages/Documenter/3fOeD/src/DocTests.jl:364
┌ Error: doctest failure in src/submodule.md:102-140
│
│ ```jldoctest
│ julia> M = FreeModule(ZZ, 2)
│ Free module of rank 2 over Integers
│
│ julia> m = M([ZZ(2), ZZ(3)])
│ (2, 3)
│
│ julia> n = M([ZZ(1), ZZ(4)])
│ (1, 4)
│
│ julia> N1, = sub(M, [m, n])
│ (Submodule over Integers with 2 generators and no relations
│ , Module homomorphism with
│ Domain: Submodule over Integers with 2 generators and no relations
│
│ Codomain: Free module of rank 2 over Integers)
│
│ julia> N2, = sub(M, [m])
│ (Submodule over Integers with 1 generator and no relations
│ , Module homomorphism with
│ Domain: Submodule over Integers with 1 generator and no relations
│
│ Codomain: Free module of rank 2 over Integers)
│
│ julia> supermodule(N1) == M
│ true
│
│ julia> iscompatible(N1, N2)
│ (true, Free module of rank 2 over Integers)
│
│ julia> issubmodule(N1, M)
│ false
│
│ julia> dim(M)
│ 2
│
│ julia> dim(N1)
│ 1
│
│ ```
│
│ Subexpression:
│
│ dim(N1)
│
│ Evaluated output:
│
│ ERROR: MethodError: no method matching dim(::AbstractAlgebra.Generic.Submodule{BigInt})
│ Closest candidates are:
│   dim(!Matched::AbstractAlgebra.Generic.YoungTableau) at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:736
│   dim(!Matched::Type{T<:Integer}, !Matched::AbstractAlgebra.Generic.YoungTableau) where T<:Integer at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:739
│   dim(!Matched::AbstractAlgebra.Generic.FreeModule{T<:Union{FieldElem, AbstractFloat, Rational}}) where T<:Union{FieldElem, AbstractFloat, Rational} at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/FreeModule.jl:44
│   ...
│ Stacktrace:
│  [1] top-level scope at none:0
│
│ Expected output:
│
│ 1
│
│   diff =
│    1ERROR: MethodError: no method matching dim(::AbstractAlgebra.Generic.Submodule{BigInt})
│    Closest candidates are:
│      dim(!Matched::AbstractAlgebra.Generic.YoungTableau) at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:736
│      dim(!Matched::Type{T<:Integer}, !Matched::AbstractAlgebra.Generic.YoungTableau) where T<:Integer at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/YoungTabs.jl:739
│      dim(!Matched::AbstractAlgebra.Generic.FreeModule{T<:Union{FieldElem, AbstractFloat, Rational}}) where T<:Union{FieldElem, AbstractFloat, Rational} at /home/martin/.local-oscar/AbstractAlgebra.jl/src/generic/FreeModule.jl:44
│      ...
│    Stacktrace:
│     [1] top-level scope at none:0
└
```

**Comment by thofma** [comment]
The `dim` functions are only provided if the type is a FieldElement, i.e. in the example one should use  `QQ` instead of `ZZ`.

**Comment by thofma** [comment]
I don't think one should swap `ZZ` with `QQ` in the whole example, but just add another small example.

**Pull request merged**